### DISCUSSION TITLE
檔案之間mprj_datal和mprj_io的轉換

### DISCUSSION THREAD

**Discussion by b3nsonchang** [bol-edu/HLS-SOC-Discussions#132] [Q&A]
RT，在fir.c中宣告要輸出的data都是用mprj_datal這個port宣告，包括start time count和end time count等等。
但在counter_la_fir_tb.v時，要checkbits時卻都是用mprj_io這個input port。
找了幾個標頭檔好像都沒看到兩者之間的聯繫，想請問是怎麼轉換的呢?

**Comment by ghost** [comment]
我想你有點誤會mprj_datal了，他不是一個port而是一個地址，你對這個地址寫值代表了wB會對這個地址發出cycle，而這個地址decode 完是mprj ，所以你的tb 的mprj port當然會吃到data

**Reply by holyuming** [reply]
想請問
1. reg_mprj_datal雖然是一個地址沒錯0x26 ...印象中。那為什麼他decode完是在mprj_io這根訊號?

2. 還是說就算是其他位置decode完也是這根訊號? (假設 WB read request 0x3000_0000這個位置，我們fir.v module會用axi-lite 回傳block level 的訊號ap_done, ap_idle ...，那這些訊號最終也會在mprj_io這根信號顯現出來嗎? 也就是整個caravel.v的某一個port: mprj_io[37:0])

**Reply by kevin861222** [reply]
小建議
1. 你可以先了解一下MMIO還有運作原理，對這個問題會有很大的幫助
2. 關於你的第一個問題，可以去參閱https://github.com/efabless/caravel/blob/main/docs/pdf/memory_map.pdf
以下是mprj_io的節錄
![截圖 2023-11-04 上午10 40 50](https://github.com/bol-edu/HLS-SOC-Discussions/assets/79128379/595f522a-3c5f-4392-b115-4f7b69dad023)

3. 關於你的第二個問題也是建議參閱memory_map，看完之後應該就能理解了

**Reply by ghost** [reply]
1. 硬體定義的
2. 你要在mprj pin顯示，第一你要確實設定好，第二要確實對這個地址給data(對硬體而言就是發WB訊號)

**Comment by boledulab** [comment]
Refer to https://youtu.be/f9zD3Fs9Bhk?feature=shared  @00:15:45 處有解釋。
